### DISCUSSION TITLE
Tips to efficiently create an independent test case for reproducing a specific exception

### DISCUSSION THREAD

**Discussion by arvindshmicrosoft** [synthetichealth/synthea#1411] [General]
I'm seeing a strange exception which I am attempting to narrow down before filing an issue. I found that for a very specific case, a call to person.getVitalSign(VitalSign.SYSTOLIC_BLOOD_PRESSURE, time) [here](https://github.com/synthetichealth/synthea/blob/16ee9bcb6d983c35fc4c9399fd5c464a2760029a/src/main/java/org/mitre/synthea/modules/calculators/Framingham.java#L480) is returning negative values, leading to consequent out-of-range exception in Framingham.getIndexForValueInRangelist.

To efficiently reproduce as an independent test case, I believe we need to have a `person` with identical or similar attributes, medications, etc. I say this because just re-creating a `person` with the same seed and using the same `time` as the bigger reproduction scenario, does not cause the exception in the independent test case. When I attach a debugger and try to dump out the state of the `person` at the time of the exception, it contains too many attributes, medications, etc. to efficiently reproduce inside an independent test case.

Are there any dev tips to help me narrow this down more efficiently? Is there a way to serialize the `person` class as a file and then use that in an independent test case?

If this is better suited for an Issue, do let me know. But again, here I'm trying to get an idea for developer process tips to help narrow these kinds of tricky, rare bugs down from a bigger run to an independent test case.

**Comment by dehall** [comment]
If you're able to provide steps to consistently reproduce the issue, we can definitely assist in debugging, even if it's a complex scenario.

In terms of generic advice, unfortunately I don't have much. We haven't isolated pieces of the system as well as we could have, so for instance even if the root cause of an issue is related to 1 specific module and none of the others, you can't necessarily remove those other modules because that impacts the random number sequence. At minimum you should be able to isolate the issue to a single patient, using the `-ps` flag to specify the single patient seed. For example if you normally run `./run_synthea -s 100 -p 100 ...` and the issue occurs for patient seed 12345, you could run `./run_synthea -s 100 -ps 12345 ...`

Re: serializing to a file, there are a couple options. You can use the JSONExporter which will produce something reasonably human-readable to inspect:
```
String json = org.mitre.synthea.export.JSONExporter.export(person);
```

or you can use native java serialization but this creates a binary file that can really only be parsed back in java code:

```
      try {
        FileOutputStream fos = new FileOutputStream(pathToOutput);
        ObjectOutputStream oos = new ObjectOutputStream(fos);
        oos.writeObject(person);
        oos.close();
        fos.close();
      } catch (Exception ex) {
        ...
      }

```

**Comment by arvindshmicrosoft** [comment]
Thank you @dehall for your tip to use `org.mitre.synthea.export.JSONExporter` to serialize the `person` object. Unfortunately I can't seem to distill it down further to write an independent test case. However, the good thing is that the issue is reproducible consistently under a controlled set of parameters. I logged #1413 with all the details to get some help. Thanks. I'll close this item and continue interacting on the Issue.
